assembly/buffer/index.ts

```diff
@@ -11,6 +12,42 @@ export class Buffer extends Uint8Array {
     return new Buffer(size);
   }
 
+  public static concat<T>(items: T, length: i32): Buffer {
+    if (!isArray<T>()) {
+      ERROR("Buffer.concat<T> must accept an Array where T extends Array<Uint8Array | Buffer>");
+    }
+    assert(unchecked(items[0]) instanceof Uint8Array); // Can this be a static check?
+
+    let size: usize = 0;
+    let itemCount = usize(items.length);
+    let itemsDataStart = items.dataStart;
+
+    for (let i: usize = 0; i < itemCount; i++) {
+      let item = load<usize>(itemsDataStart + (i << alignof<usize>()));
+      if (item == 0) continue;
+      size += <usize>load<u32>(item, offsetof<Uint8Array>("dataLength"));
+    }
+    size = min<usize>(usize(length), size);
+    assert(items);
+
+    let buffer = __alloc(size, idof<ArrayBuffer>());
+    let result = changetype<Buffer>(__alloc(offsetof<Buffer>(), idof<Buffer>()));
+
+    result.data = changetype<ArrayBuffer>(buffer);
+    result.dataStart = changetype<usize>(buffer);
+    let start: usize = result.dataStart;
+    for (let i: usize = 0; i < itemCount && size > 0; i++) {
+      let item = load<usize>(itemsDataStart + (i << alignof<usize>()));
+      if (item == 0) continue;
+      let count = min<u32>(size, <usize>load<u32>(item, offsetof<Uint8Array>("dataLength")));
+      memory.copy(start, load<usize>(item, offsetof<Uint8Array>("dataStart")), count);
+      start += count;
+      size -= count;
+    }
+
+    return result;
+  }
+
   @unsafe public static allocUnsafe(size: i32): Buffer {
```

Comment:
I *think* I'm fine with annotating this `@unsafe`, just wanting to note that it isn't actually unsafe in a sense that affects the runtime because memory never overlaps and as such can't mess with RT structures. It's of course unsafe in a sense that untrusted code can call this repeatedly to recover previous data from memory.

Reply:
Btw, one nit I missed before was that the changetype on `result.dataStart = changetype<usize>(buffer); below` is unnecessary since `buffer` is already `usize`. Adding this comment here since I can't add it below :)

Reply:
I'll fix that in this pull request. Feel free to Comment on anything that could be made better. I want to be very proud of this work :)

Reply:
This nit should be all set with the latest commit.